hls4ml/converters/pytorch_to_hls.py

```diff
@@ -18,17 +18,24 @@ def get_weights_data(self, layer_name, var_name):
         data = None
 
         # Workaround for naming schme in nn.Sequential,
-        # have to remove the prefix we previously had to add to make sure the tensors are found
-        if 'layer_' in layer_name:
-            layer_name = layer_name.split('layer_')[-1]
-
-        # if a layer is reused in the model, torch.FX will append a "_n" for the n-th use
-        # have to snap that off to find the tensors
-        if layer_name.split('_')[-1].isdigit() and len(layer_name.split('_')) > 1:
-            layer_name = '_'.join(layer_name.split('_')[:-1])
-
-        if layer_name + '.' + var_name in self.state_dict:
-            data = self.state_dict[layer_name + '.' + var_name].numpy()
+        # We have to remove the prefix we added earlier to not have layer
+        # names start with an underscore. However, this clashes with cases
+        # were an named nn.Sequential with "layer" in it's name is used.
+        # We make sure that we pick the correct name for the tensor
+        layerInKeyName = False
+        for key in self.state_dict.keys():
+            if "layer." in key:
+                layerInKeyName = True
+
+        if '_' in layer_name:
```

Comment:
I am not sure I follow the logic here, but I am not a pytorch expert. Can you double-check?

Reply:
I had another look and this is a bit convoluted, but it works. The issue is that the layers inside a torch.nn.Sequential get named in the pattern `nameOfSquential_n` where `n` just numbers the layers inside the sequential. If the sequential does not have a name, which happens if you just go `model = nn.Sequential(..)`, the layers will have names that just start with an underscore, which hls4ml doesn't like. So we add the prefix "layer" to those in the loop over the layers, which we have to remove when we go and load the tensors. The changes in this PR account for the fact that someone could create a named nn.Sequential just named `layer`, which then clashes with our previous assumption that if a layer name starts with `layer_` it is because we added it by hand to get around the issue with layer names starting with just an `_`.

A further complication is that while torch.FX reports these structured layer names with underscores while in the `state_dict` a `.` is used, so we have to replace the `_` with `.` here. The last complication is the case of layers that used multiple times in the same model. torch.FX reports them as different layers, adding an `_n` for the n-th time it is used to the layer name. But the tensors in the state dict will not have those modifications, so we also have to account for this. 

Bit of a mess, but I have tested all these cases and this implementations catches all edge cases that I'm aware off.

Reply:
This has now been solved significantly nicer by Vladimir :)